internal/gemalto/key-secure.go

```diff
@@ -0,0 +1,354 @@
+	Domain string        // The KeySecure domain - similar to a Vault Namespace
+	Retry  time.Duration // The time to wait before trying to re-authenticate
+}
+
+// KeySecure is a Gemalto KeySecure client that
+// stores / fetches key-value pairs as secrets.
+//
+// It tries to connect to a KeySecure instance
+// at the given endpoint and uses the login
+// credentials to authenticate.
+type KeySecure struct {
+	// Endpoint is the KeySecure instance endpoint.
+	Endpoint string
+
+	// CAPath is a path to the root CA certificate(s)
+	// used to verify the TLS certificate of the KeySecure
+	// instance. If empty, the host's root CA set is used.
+	CAPath string
+
+	// Login credentials are used to authenticate to the
+	// KeySecure instance and obtain a short-lived authentication
+	// token.
+	Login Credentials
+
+	// ErrorLog specifies an optional logger for errors.
+	// If an unexpected error is encountered while trying
+	// to fetch, store or delete a key or when an authentication
+	// error happens then an error event is written to the error
+	// log.
+	//
+	// If nil, logging is done via the log package's standard
+	// logger.
+	ErrorLog *log.Logger
+
+	client *client
+}
+
+// Authenticate tries to establish a connection to a
+// KeySecure server using the login credentials.
+//
+// It retruns an error if no connection could be
+// established - for instance because of invalid
+// credentials.
+func (s *KeySecure) Authenticate() (err error) {
+	var rootCAs *x509.CertPool
+	if s.CAPath != "" {
+		rootCAs, err = loadCustomCAs(s.CAPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	s.client = &client{
+		ErrorLog: s.ErrorLog,
+		Retry: xhttp.Retry{
+			Client: http.Client{
+				Transport: &http.Transport{
+					TLSClientConfig: &tls.Config{
+						RootCAs: rootCAs,
+					},
+					Proxy: http.ProxyFromEnvironment,
+					DialContext: (&net.Dialer{
+						Timeout:   10 * time.Second,
+						KeepAlive: 10 * time.Second,
+						DualStack: true,
+					}).DialContext,
+					ForceAttemptHTTP2:     true,
+					MaxIdleConns:          100,
+					IdleConnTimeout:       30 * time.Second,
+					TLSHandshakeTimeout:   10 * time.Second,
+					ExpectContinueTimeout: 1 * time.Second,
+				},
+			},
+		},
+	}
+	if err = s.client.Authenticate(s.Endpoint, s.Login); err != nil {
+		return err
+	}
+	go s.client.RenewAuthToken(context.Background(), s.Endpoint, s.Login)
+	return nil
+}
+
+// Create creates the given key-value pair at Vault if and only
+// if the given key does not exist. If such an entry already exists
+// it returns kes.ErrKeyExists.
+func (s *KeySecure) Create(key, value string) error {
+	type Request struct {
+		Type  string `json:"dataType"`
+		Value string `json:"material"`
+		Name  string `json:"name"`
+	}
+
+	body, err := json.Marshal(Request{
+		Type:  "seed", // KeySecure supports blob, password and seed
+		Value: value,
+		Name:  key,
+	})
+	if err != nil {
+		return err
+	}
+
+	url := fmt.Sprintf("%s/api/v1/vault/secrets", s.Endpoint)
```

Comment:
should this be gemalto? or Gemalto uses vault under the hood?

Reply:
No, Gemalto KeySecure and Hashicorp Vault are two different (KMS) vendors.